jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -503,130 +509,40 @@ public static String parentPath(String p)
      */
     public static String canonicalPath(String path)
     {
-        if (path==null || path.length()==0)
+        if (path == null || path.isEmpty() || !path.contains("."))
             return path;
 
-        int end=path.length();
-        int start = path.lastIndexOf('/', end);
+        if(path.startsWith("/.."))
+            return null;
 
-    search:
-        while (end>0)
-        {
-            switch(end-start)
-            {
-              case 2: // possible single dot
-                  if (path.charAt(start+1)!='.')
-                      break;
-                  break search;
-              case 3: // possible double dot
-                  if (path.charAt(start+1)!='.' || path.charAt(start+2)!='.')
-                      break;
-                  break search;
-            }
-            
-            end=start;
-            start=path.lastIndexOf('/',end-1);
-        }
+        List<String> directories = new LinkedList<>();
```

Comment:
Just to explain my concern,   linked list creates lots of little objects, which when unlinked will need to be garbage collected.   While they are typically trivial to collect and may add perhaps no time to the collection algorithm itself, the fact remains that the more garbage generated, then the more frequent we need to collect, which may mean that more objects move out of the eden space and eventually into perm gen.  Such object can then be very expensive to collect.

I'm garbage phobic through long experience...